core/internal/stream/reader.go

```diff
@@ -0,0 +1,92 @@
+package stream
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/wandb/wandb/core/internal/observability"
+	"github.com/wandb/wandb/core/internal/runwork"
+	"github.com/wandb/wandb/core/internal/settings"
+	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
+	"google.golang.org/protobuf/proto"
+)
+
+type ReaderParams struct {
+	Logger   *observability.CoreLogger
+	Settings *settings.Settings
+	RunWork  runwork.RunWork
+}
+
+type Reader struct {
+	settings *settings.Settings
+	logger   *observability.CoreLogger
+	store    *Store
+	runWork  runwork.RunWork
+}
+
+func NewReader(params ReaderParams) *Reader {
+	return &Reader{
+		settings: params.Settings,
+		logger:   params.Logger,
+		runWork:  params.RunWork,
+	}
+}
+
+func (r *Reader) Do() {
+	defer r.logger.Reraise()
+	r.logger.Info("reader: Do: started", "stream_id", r.settings.GetRunID())
+
+	r.store = NewStore(r.settings.GetTransactionLogPath())
```

Comment:
Add a TODO: to track/report read progress?

Reply:
We can scan the whole file to check the number of records, check whether a Finish record is present and so on.